### ISSUE TITLE
Interrupt handler registration and execution 

### ISSUE THREAD

**Issue by cahirwpz** [cahirwpz/mimiker#32]
Make possible for any part of kernel to register an interrupt handler. An interrupt handler returns true when the handler processed the interrupt. For reference look at [sys/interrupt.h](http://fxr.watson.org/fxr/source/sys/interrupt.h), [kern/kern_intr.c](http://fxr.watson.org/fxr/source/kern/kern_intr.c) and [sys/bus.h](http://fxr.watson.org/fxr/source/sys/bus.h).

```
#define FILTER_STRAY   0x01 /* this device did not trigger the interrupt */
#define FILTER_HANDLED 0x02 /* the interrupt has been fully handled and can be EOId */

/*
 * The filter routine is run in primary interrupt context and may not
 * block or use regular mutexes.  The filter may either completely
 * handle the interrupt or it may perform some of the work and
 * defer more expensive work to the regular interrupt handler.
 */
typedef int driver_filter_t(void*);
typedef void driver_intr_t(void*);

/*
 * Describe a hardware interrupt handler.
 * Multiple interrupt handlers for a specific event can be chained together.
 */
typedef struct intr_handler {
    TAILQ_ENTRY(intr_handler) ih_next; /* Next handler for this event. */
    driver_filter_t *ih_filter;        /* Filter handler function. */
    driver_intr_t *ih_handler;         /* Handler function. */
    intr_event_t *ih_event;            /* Event we are connected to. */
    void *ih_argument;                 /* Argument to pass to handlers. */
    char *ih_name;                     /* Name of handler. */
    prio_t ih_prio;                    /* Priority of this handler. */
} intr_handler_t;

/*
 * Describe an interrupt event.
 * An event holds a list of handlers.
 */
typedef struct intr_event {
    TAILQ_ENTRY(intr_event) ie_list;
    TAILQ_HEAD(, intr_handler) ie_handlers; /* Interrupt handlers. */
    char *ie_name;                          /* Individual event name. */
    uint8_t ie_irq;                         /* Physical irq number. */
    ...
} intr_event_t;

void intr_init();
void intr_event_init(intr_event_t *ie, uint8_t irq, char *name)
void intr_event_add_handler(intr_event_t *ie, intr_handler_t *ih);   
void intr_event_remove_handler(intr_handler_t *ih);
void intr_event_execute_handlers(intr_event_t *ie);
```

**Comment by coodie** [comment]
Before i start digging the topic. Do you know if this implementation on wifire would differ greatly from the malta we'd like to port it to? I'd like to know if there is a point in doing this, before i (might) realise i just wasted time.

**Comment by cahirwpz** [comment]
Qemu doesn't implement vectored interrupts with non-zero offsets. Thus we configure interrupts in almost the same way as on original MIPS platforms. There're some minor differences but I strongly believe the rest will be the same.

**Comment by coodie** [comment]
Keep in mind that don't understand problem well, that's why i'm starting discussion.

What is purpose of filter handler function? As far as I understood `freebsd` code, filter function is the first one to be executed when interrupt comes, it's name is filter, because some interrupts are stray (what is stray interrupt and when can it occur ?), and it filters them in case. Filter is executed in interrupt routine so this cannot use some complicated synchronisation methods (I don't know why, but I have some intuition), that's why some work has to be deferred to threads which can basically do everything. A little bit about this here:

https://lists.freebsd.org/pipermail/freebsd-hackers/2009-January/027552.html

Since `freebsd` uses special threads to handle interrupts (and we don't have threads yet), is there a point in making this task now, before we have threads? Without threads we'll have to restrict ourselves from using some functionality (which doesn't exist at the moment, but it will).

What kinds of interrupt would you want to handle right now? Not every interrupt is IRQ based (like exceptions for example), there is 190 of them on pic32mz, and we mostly care about clock interrupt at the moment and maybe software interrupts (traps) to make system calls. If we plan to use some simple code in interrupt handlers, then we might not need to use threads to implement this.

**Comment by cahirwpz** [comment]
I'll consider this task to be finished when at least two devices connected to single interrupt are correctly handled. To achieve it in the simplest form you should program `RTC` and `UART0` to generate interrupts: https://github.com/cahirwpz/mimiker/blob/master/include/malta.h#L38

**Comment by cahirwpz** [comment]
@dyniec There's already an implementation of `intr_handler` and `intr_chain` machine independent functions in `sys/interrupt.c`. Please read `bus_setup_intr(9)` manual.

**Comment by cahirwpz** [comment]
Implemented with 2126c0ff04c3c1e29a23b76db5e46770a9d2024c
